libraries/core_libs/consensus/src/vote_manager/vote_manager.cpp

```diff
@@ -47,35 +47,37 @@ void VoteManager::retreieveVotes_() {
   }
 }
 
-bool VoteManager::addUnverifiedVote(std::shared_ptr<Vote> const& vote) {
+bool VoteManager::addUnverifiedVote(std::shared_ptr<Vote>&& vote) {
   uint64_t pbft_round = vote->getRound();
   const auto& hash = vote->getHash();
   vote->getVoterAddr();  // this will cache object variables - speed up
 
+  if (voteInUnverifiedMap(pbft_round, hash)) {
+    LOG(log_dg_) << "Vote " << hash << " is already in unverified map";
+    return false;
+  }
```

Comment:
Why need check voteInUnverifiedMap first? What's the benefits?

Reply:
It might happen quite often that you receive the same vote from multiple nodes and because we are processing votes concurrently, you want to eliminate "unique" locking as much as possible because it blocks the other threads. By checking voteInUnverifiedMap first, you use shared_lock, which does not block other threads - this might improve performance a lot